src/modsim2/similarity/metrics/kde.py

```diff
@@ -0,0 +1,295 @@
+            labels_A: The target values for the first dataset
+            labels_B: The target values for the second dataset
+            embedding_name: The feature embeddings to use for the input arrays
+            kernel_name: The kernel to be applied in the density estimation, any
+                    kernels allowed in sklearns KernelDensity class are valid.
+                    Currently these are 'gaussian', 'tophat', 'epanechnikov',
+                    'exponential', 'linear', 'cosine'
+            metric_name: The metric to be used to calculate the distance between
+                    the density functions, currently only 'l2' is valid
+            embedding_kwargs: A dictionary of key word arguments to be passed
+                    to the embedding function
+            integration_kwargs: A dictionary of key word arguments to be passed
+                    to the integration function
+        """
+        # A value of zero is returned if the datasets are the same (the integration
+        # methods cannot be used if the datasets are the same, and the approximate
+        # kl method will calculate as -inf)
+        if np.array_equal(data_A, data_B):
+            return 0
+
+        # Check for valid embedding choice
+        assert embedding_name in EMBEDDING_FN_DICT, "Error: embedding does not exist"
+
+        if embedding_name in ["inception_umap"]:
+            # as UMAP is stochastic, a random seed is required
+            # for reproducability
+            embed_random_seed = randint(low=0, high=100, size=(1,))[0]
+            embedding_kwargs["random_seed"] = int(embed_random_seed)
+
+        # Embed the data
+        embed_A, embed_B = self._embed_data(
+            data_A=data_A,
+            data_B=data_B,
+            embedding_name=embedding_name,
+            embedding_kwargs=embedding_kwargs,
+        )
+
+        if metric_name == "kl_approx":
+            # The kernel density estimation is not required
+            distance_AB = self.kl_tree(embed_A, embed_B)
+            distance_BA = self.kl_tree(embed_B, embed_A)
+            return [distance_AB, distance_BA]
+
+        else:
+            # Create kernel density estimators for A & B
+            # Scott's rule estimates an appropriate value for the bandwidth, an
+            # alternative would be to search bandwidth values using CV
+            estimator_A = KernelDensity(kernel=kernel_name, bandwidth="scott")
+            estimator_A.fit(embed_A)
+            estimator_B = KernelDensity(kernel=kernel_name, bandwidth="scott")
+            estimator_B.fit(embed_B)
+
+            # Confirm the two embedded datasets have the same number of features
+            # before setting the number of features
+            assert (
+                embed_A.shape[1] == embed_B.shape[1]
+            ), "Error: datasets A & B do not have the same number of features"
+            num_features = embed_A.shape[1]
+
+            if metric_name == "l2":
```

Comment:
ideally all should return two values if any of them do

Reply:
Unfortunately can't be done - KL divergence differs from other similarity measures in that it's not a metric (i.e. it doesn't satisfy d(A,B) =/= d(B,A)), hence the need to return two measures. In the same manner it's not possible for the others to return two values. Could structure the KL values in a single object, but this would still be different in nature from what the others return. Can discuss more over a call

Reply:
Following yesterday's discussion, my above comment is to be ignored and will be switching to two values for all metrics (just returning same twice)

Have added a dictionary lookup for the methods

Reply:
All of these metrics now return 2 values. Have added a new comment below to do this for other metrics